js/diff.js

```diff
@@ -0,0 +1,150 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+class Diff {
+    constructor() {
+        this.withNewChildren = new Map()
+        this.withRemovedChildren = new Map()
+        this.withMovedBoxPos = []
+        this.withReorderedChildren = []
+    }
+
+    markNewChild(parent, child) {
+        let value = this.withNewChildren.get(parent)
+        if (value == null) {
+            value = []
+        }
+        value.push(child)
+        this.withNewChildren.set(parent, value)
```

Comment:
I don't think this should go inside the if-condition, because if the lookup is not null (which would be the case if two views were added to a ViewParent in the same frame), we still want to include both of them in the array.